src/storage/src/source/postgres/snapshot.rs

```diff
@@ -270,17 +282,34 @@ pub(crate) fn render<G: Scope<Timestamp = MzOffset>>(
                 let client = connection_config
                     .connect_replication(&config.config.connection_context.ssh_tunnel_manager)
                     .await?;
-                // The main slot must be created *before* we start snapshotting so that we can be
-                // certain that the temporarly slot created for the snapshot start at an LSN that
-                // is greater than or equal to that of the main slot.
-                super::ensure_replication_slot(&client, &connection.publication_details.slot)
-                    .await?;
 
-                let snapshot_info = export_snapshot(&client).await?;
+                // Attempt to export the snapshot by creating the main replication slot. If that
+                // succeeds then there is no need for creating additional temporary slots.
+                let main_slot = &connection.publication_details.slot;
+                let snapshot_info = match export_snapshot(&client, main_slot, false).await {
+                    Ok(info) => info,
+                    Err(err @ TransientError::ReplicationSlotAlreadyExists) => {
+                        match connection.connection.flavor {
+                            // If we're connecting to a vanilla we have the option of exporting a
+                            // snapshot via a temporary slot
+                            PostgresFlavor::Vanilla => {
+                                let tmp_slot = format!(
+                                    "mzsnapshot_{}",
+                                    uuid::Uuid::new_v4()).replace('-', ""
+                                );
+                                export_snapshot(&client, &tmp_slot, true).await?
+                            }
+                            // No salvation for Yugabyte
+                            PostgresFlavor::Yugabyte => return Err(err),
```

Comment:
does this mean yugabyte sources will always fail if we need to re-snapshot when a new source export is added, since the main replication slot will already exist?

Reply:
Yes, that's the current limitation until Yugabyte offers us a few more APIs that should land in early 2025